### ISSUE TITLE
Feature Request: Wrap b2Settings.h definitions with #ifndef

### ISSUE THREAD

**Issue by MathewMacDougall** [erincatto/box2d#559]
It would be useful for developers to have more control over the variables defined in [b2Settings.h](https://github.com/erincatto/Box2D/blob/master/Box2D/Common/b2Settings.h) in order to tailor the physics behavior for their application. It would also be preferable to be able to do this without having to fork and modify Box2D.

### Example Use Case
I am writing a simulator for soccer-playing robots, and need collisions to be elastic even at low speeds. The default value of `b2_velocityThreshold` doesn't work in this case because all collisions happening below 1m/s will "stick".

Therefore I'd like to be able to change that value, preferably without forking the Box2D repository so that there are fewer things to manage in the project.

### Proposed Solution
Wrap all (or most) of the defined variables in [b2Settings.h](https://github.com/erincatto/Box2D/blob/master/Box2D/Common/b2Settings.h) with `#ifndef` blocks. That way developers using the library can define the variables themselves if they want to change them, but behavior is otherwise unchanged.

[See here for an example.](https://github.com/UBC-Thunderbots/Box2D/commit/cba3121823241634638e2b294615df6c157cf4be)

With this change I can manually define the variables with the values I want directly within my project and build-system. 

Here's an example bazel rule I'm using in my project with this new `#ifndef` system.
```
cc_library(
    name = "box2d",
    srcs = glob(["Box2D/**/*.cpp"]),
    hdrs = glob(["Box2D/**/*.h"]),
    includes = [
        ".",
    ],
    defines = [
        "b2_maxPolygonVertices=10",
        "b2_linearSlop=0.001f",
        "b2_velocityThreshold=0.001f",
    ],
    visibility = ["//visibility:public"],
)
```

**Comment by erincatto** [comment]
What I've done in other projects is force the user to include a file not provided by Box2D. Like this
b2Settings.h:
```
#ifdef b2TESTBED
...
#define b2_maxManifoldPoints	2
...
#else
#include "b2UserSettings.h"
#endif
```
This has the ability to provide user data injection as well. For example instead of b2Body holding a void* it could hold a b2BodyUserData structure. This might be a bit onerous for most users. So I could flip it around and only include b2UserSettings.h if b2USER_SETTINGS is defined.

What do you think?

**Comment by MathewMacDougall** [comment]
I think that's a good idea, and the user data injection is a nice bonus :+1:

**Comment by erincatto** [comment]
Take a look at https://github.com/erincatto/box2d/pull/643

**Comment by AshleyScirra** [comment]
Is #643 a built-time only feature? If so it does not fix #601, which requests the option is made runtime-configurable. That issue was marked a duplicate of this issue, even though this issue appears to be solvable with build-time options only, but the issue I filed is not solvable with build-time options only.

**Comment by erincatto** [comment]
True, I can probably make that particular value run-time tunable.

**Comment by erincatto** [comment]
Fixed in #643
